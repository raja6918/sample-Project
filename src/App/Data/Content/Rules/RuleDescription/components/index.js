import RuleTableDialogWithConnect from './RuleTableDialog/RuleTableDialog';

export { RuleTableDialogWithConnect };
