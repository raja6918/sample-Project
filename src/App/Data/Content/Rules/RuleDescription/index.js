import RuleDescriptionWithTranslation from './RuleDescription';

export default RuleDescriptionWithTranslation;
