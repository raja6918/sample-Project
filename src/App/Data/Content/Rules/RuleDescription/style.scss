.rule-description {
    position: relative;
    background-color: #f0f8ff;
    color: #000000;
    white-space: pre-wrap;
    font-size: 13px;

    .overlay {
        position: absolute;
        top: 0;
        left: 0;
        right: 0;
        bottom: 0;
        background-color: rgba(240, 248, 255, 0.8);
        width: 100%; 
        height: 100%;
        z-index: 1000;
        cursor: pointer;
    }

    & .MuiFormControl-root {
        vertical-align: unset;
    }
    & .MuiInput-input {
        font-size: 13px;
        font-weight: 600;
        color: #ff650c;
    }
    & .MuiSelect-selectMenu {
        font-weight: 400;
    }
    & .MuiInput-underline:hover:before {
        border-bottom-color: #ff650c;
    }
    & .MuiInput-underline:before,
    .MuiInput-underline:after {
        border-bottom-color: #ff650c;
    }
    & .MuiSelect-icon{
        color: #ff650c;
    }
    & .input-error .MuiInput-input {
        color: #000000;
    }

    & .input-error .MuiInput-underline:hover:before {
        border-bottom-color: #d50000;
    }

    & .input-error {
        .MuiInput-underline:before,
        .MuiInput-underline:after {
            border-bottom-color: #d50000;
        }
        & .MuiSelect-icon{
            color: #d50000;
        }
    } 

    & .date-time-pointer .MuiInput-input{
        cursor: pointer;
    }

    & .date-time-pointer .Mui-disabled{
        cursor: default;
    }

    & .MuiInputBase-root.Mui-disabled {
        & .MuiInput-input {
            color: rgba(0, 0, 0, 0.26); 
        }  
        & .MuiSelect-icon {
            color: rgba(0, 0, 0, 0.26);
        }
    }

    & .MuiInput-underline.Mui-disabled:hover:before {
        border-bottom-color: rgba(0, 0, 0, 0.26); 
    }
    
    & .MuiInput-underline.Mui-disabled::before,
    .MuiInput-underline.Mui-disabled::after {
        border-bottom-color: rgba(0, 0, 0, 0.26); 
    }

    label[class*="focus"] + div:after, div[class*=focus]::after {
        background-color: transparent !important; 
    }
}