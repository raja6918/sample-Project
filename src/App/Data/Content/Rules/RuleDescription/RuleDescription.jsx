import React, { Component, Fragment, createRef } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { translate } from 'react-i18next';
import DescriptionWithConnect from './Description';
import CloseIcon from '@material-ui/icons/Close';
import IconButton from '@material-ui/core/IconButton';
import withErrorHandler from '../../../../../components/ErrorHandler/withErrorHandler';
import { checkElementVisibility } from '../../../../../utils/common';
import isEmpty from 'lodash/isEmpty';
import './style.scss';

const PrimarySessionDiv = styled.div`
  padding: 14px 56px 14px 56px;
`;

const SecondarySessionDiv = styled.div`
  padding: 0px 56px 0px 56px;
`;

const SubSessionDiv = styled.div`
  display: flex;
  padding: 14px 0px 14px 0px;
`;

const BorderDiv = styled.div`
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
`;

const SubSessionsTitle = styled.div`
  color: #000000;
  font-weight: bold;
  font-size: 13px;
  line-height: 17px;
`;

export class RuleDescription extends Component {
  state = {
    subSessions: [],
    overlay: false,
  };

  containerRef = createRef();

  previouslyMovedUpRule = null;
  selectedLink = null;

  componentDidMount() {
    // Fetch parent description
    this.props.fetchDescription(this.props.data);
  }

  getDescription = (code, ruleDescriptions) => {
    const ruleDescription = ruleDescriptions.find(
      ruleDescription => ruleDescription.code === code
    );

    return ruleDescription ? ruleDescription : null;
  };

  handleLink = (data, ruleSet) => {
    const checkExit = this.state.subSessions.find(code => code === data.code);

    if (!checkExit) {
      this.selectedLink = data.code;
      this.setState(
        prevState => ({ subSessions: [...prevState.subSessions, data.code] }),
        () => {
          this.props.fetchDescription(data, ruleSet);
          // Manually calling columsResizing method in GenericCollapsibleTable to fix any size issue caused by detail session
          this.props.triggerResize();
        }
      );
    }
  };

  /**
   * To fix bug with new scrollbar ie unwanted space at bottom of table body when last row's subsessions are collapsed.
   * Here we calculate subsession height that we need to move up the scrollbar to fix the spacing issues.
   */
  calculateMoveUpHeight = subSessionRef => {
    try {
      const height = subSessionRef.current
        ? subSessionRef.current.clientHeight
        : 0;
      const moveUpHeight =
        this.props.scrollRefY.clientHeight - height < 0
          ? this.props.scrollRefY.clientHeight
          : height;

      return {
        moveUpHeight: moveUpHeight + 40,
      };
    } catch (error) {
      console.error(error);
      return {
        moveUpHeight: 0,
      };
    }
  };

  handleClose = (code, subSessionRef) => {
    const { scrollRefY, needToMoveUp } = this.props;
    // To fix bug with new scrollbar ie unwanted space at bottom of table body when last row's subsession collapsed.
    if (scrollRefY && needToMoveUp) {
      const { moveUpHeight } = this.calculateMoveUpHeight(subSessionRef);
      const currentPos = scrollRefY.scrollTop;
      scrollRefY.scrollTop -= moveUpHeight;
      /**
       * For ALT-2311 - When an expanded rule located toward the bottom is contracted,
       * either by collapsing the rule description altogether or by closing an extension section,
       * the rule remains at the same vertical height (i.e., the top edge of the row stays in place)
       * So here we reset back to original position.
       */
      const timer = setTimeout(() => {
        scrollRefY.scrollTop = currentPos;
        clearTimeout(timer);
      });
    }

    this.setState(
      prevState => ({
        subSessions: prevState.subSessions.filter(data => data !== code),
      }),
      () => {
        // Manually calling columsResizing method in GenericCollapsibleTable to fix any size issue caused by detail session
        this.props.triggerResize();
      }
    );
  };

  setOverlay = () => {
    this.setState({ overlay: true });
  };

  removeOverlay = () => {
    this.props.clearErrorNotification();
    this.setState({ overlay: false });
  };

  reportError = error => {
    this.props.reportError({ error });
  };

  /**
   * For ALT-2311 - When a rule located toward the bottom of the Rules table is expanded, the view is adjusted
   * (vertical scroll) such that the rule expansion section is entirely visible in the page.
   * And When an expanded rule located toward the bottom of the Rules table is extended by clicking a link in the
   * rule description, the view is adjusted (vertical scroll) such that the rule expansion section is entirely visible in the page.
   */
  moveUp = (elementRef, delay = 500) => {
    const timer = setTimeout(() => {
      try {
        /**
         * Giving a samll delay to make sure that element is properly rendered to get its height.
         * Go with setTimeout approach intead of calling this method from componentDidMount and ComponentDidUpdate because
         * the scrollbar move up is not working in the last row (ie scrollRefY.scrollTop is not mutating) without giving 500ms delay.
         */
        const { scrollRefY } = this.props;
        const { isVisible, bottomHeightDiff } = checkElementVisibility(
          scrollRefY,
          elementRef.current
        );
        if (scrollRefY && !isVisible && bottomHeightDiff > 0) {
          scrollRefY.scrollTop += bottomHeightDiff;
        }
        clearTimeout(timer);
      } catch (error) {
        console.error(error);
      }
    }, delay);
  };

  render() {
    const { subSessions, overlay } = this.state;
    const {
      t,
      data,
      userId,
      readOnly,
      scenarioId,
      ruleDescriptions,
      handleParamSet,
      handleParamReset,
      toggleLoader,
      isOpen,
      lastOpenedRowData,
      refreshRules,
    } = this.props;

    const description = this.getDescription(data.code, ruleDescriptions);
    const parentDescriptionCode = data.code;

    const checkCurrentRowOpenedWithData =
      isOpen &&
      lastOpenedRowData &&
      lastOpenedRowData.code === data.code &&
      description;
    const checkSubMenusNotOpen =
      Array.isArray(subSessions) && subSessions.length === 0;
    const checkDuplicateMovedUp = this.previouslyMovedUpRule !== data.code;

    if (
      checkCurrentRowOpenedWithData &&
      checkSubMenusNotOpen &&
      checkDuplicateMovedUp
    ) {
      this.moveUp(this.containerRef, 500);
      this.previouslyMovedUpRule = data.code;
    }

    return (
      <div className="rule-description" ref={this.containerRef}>
        {overlay && <div className="overlay" onClick={this.removeOverlay} />}
        {/**
         * The rendering primary session
         */}
        <PrimarySessionDiv>
          {description && description.userDescription && (
            <DescriptionWithConnect
              t={t}
              code={parentDescriptionCode}
              data={description.userDescription}
              ruleSet={description.ruleset}
              handleParamSet={handleParamSet}
              handleParamReset={handleParamReset}
              handleLink={this.handleLink}
              overlay={overlay}
              setOverlay={this.setOverlay}
              removeOverlay={this.removeOverlay}
              userId={userId}
              scenarioId={scenarioId}
              readOnly={readOnly}
              toggleLoader={toggleLoader}
              reportError={this.reportError}
              refreshRules={refreshRules}
            />
          )}
        </PrimarySessionDiv>
        {/**
         * The rendering of subsessions by looping subSessions state
         */}
        {Array.isArray(subSessions) && subSessions.length > 0 && (
          <SecondarySessionDiv>
            {subSessions.length > 0 && <BorderDiv />}
            {subSessions.map(code => {
              const description = this.getDescription(code, ruleDescriptions);
              const subSessionRef = createRef();

              if (description === null) {
                return <Fragment key={code} />;
              }

              if (this.selectedLink === code) {
                this.moveUp(this.containerRef, 500);
                this.selectedLink = null;
              }

              return (
                <SubSessionDiv key={code}>
                  <span ref={subSessionRef}>
                    <SubSessionsTitle
                      style={{
                        paddingBottom: isEmpty(
                          description.userDescription.references
                        )
                          ? '4px'
                          : '',
                      }}
                    >
                      {description.name}
                    </SubSessionsTitle>
                    {description && description.userDescription && (
                      <DescriptionWithConnect
                        t={t}
                        code={code}
                        data={description.userDescription}
                        ruleSet={description.ruleset}
                        handleParamSet={handleParamSet}
                        handleParamReset={handleParamReset}
                        handleLink={this.handleLink}
                        overlay={overlay}
                        setOverlay={this.setOverlay}
                        removeOverlay={this.removeOverlay}
                        userId={userId}
                        scenarioId={scenarioId}
                        readOnly={readOnly}
                        toggleLoader={toggleLoader}
                        reportError={this.reportError}
                        refreshRules={refreshRules}
                      />
                    )}
                  </span>
                  <span style={{ flex: 1 }}>
                    <IconButton
                      style={{ float: 'right' }}
                      aria-label="close row"
                      size="small"
                      onClick={() => this.handleClose(code, subSessionRef)}
                    >
                      <CloseIcon color="disabled" fontSize="small" />
                    </IconButton>
                  </span>
                </SubSessionDiv>
              );
            })}
          </SecondarySessionDiv>
        )}
      </div>
    );
  }
}

RuleDescription.propTypes = {
  t: PropTypes.func.isRequired,
  data: PropTypes.shape().isRequired,
  ruleDescriptions: PropTypes.arrayOf(PropTypes.shape()).isRequired,
  handleParamSet: PropTypes.func.isRequired,
  handleParamReset: PropTypes.func.isRequired,
  fetchDescription: PropTypes.func.isRequired,
  triggerResize: PropTypes.func.isRequired,
  scrollRefY: PropTypes.shape({
    clientHeight: PropTypes.number,
    scrollTop: PropTypes.number,
  }).isRequired,
  needToMoveUp: PropTypes.bool.isRequired,
  userId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  scenarioId: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
    .isRequired,
  readOnly: PropTypes.bool.isRequired,
  toggleLoader: PropTypes.func.isRequired,
  reportError: PropTypes.func.isRequired,
  clearErrorNotification: PropTypes.func.isRequired,
  isOpen: PropTypes.bool.isRequired,
  lastOpenedRowData: PropTypes.shape({ code: PropTypes.string.isRequired }),
  refreshRules: PropTypes.func.isRequired,
};

RuleDescription.defaultProps = {
  lastOpenedRowData: null,
};

export default translate()(withErrorHandler(RuleDescription, 360000));
