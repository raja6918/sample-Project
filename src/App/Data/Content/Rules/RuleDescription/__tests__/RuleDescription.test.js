import React from 'react';
import { shallow } from 'enzyme';
import { RuleDescription } from '../RuleDescription';
import { ruleDescriptionData, rulesData } from '../../mockData';

describe('Test Rule Description Component', () => {
  const minProps = {
    t: jest.fn(msg => msg),
    handleParamSet: jest.fn(),
    handleParamReset: jest.fn(),
    fetchDescription: jest.fn(),
    triggerResize: jest.fn(),
    clearErrorNotification: jest.fn(),
    code: 'CiesApplyAllowDeadheadDuty',
    ruleDescriptions: ruleDescriptionData,
    data: rulesData[0],
    scrollRefY: {
      clientHeight: 500,
      scrollTop: 0,
    },
    needToMoveUp: true,
    userId: 1,
    scenarioId: 1,
    readOnly: false,
    toggleLoader: jest.fn(),
    reportError: jest.fn(),
    isOpen: true,
    lastOpenedRowData: rulesData[0],
    refreshRules: jest.fn(),
  };

  let wrapper;
  beforeAll(() => {
    wrapper = shallow(<RuleDescription {...minProps} />);
  });

  test('Rule Description Component renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('Rule Description subSessions state and subsession renders correctly when handleLink is invoked', () => {
    wrapper
      .find('Description')
      .at(0)
      .prop('handleLink')({ code: minProps.code });

    expect(wrapper.state('subSessions')).toEqual([minProps.code]);
    expect(minProps.fetchDescription).toBeCalled();
    expect(minProps.triggerResize).toBeCalled();
    expect(wrapper).toMatchSnapshot();
  });

  test('Rule Description subSessions state reset and subsession renders correctly when handleClose is invoked', () => {
    wrapper.find({ 'aria-label': 'close row' }).prop('onClick')();

    expect(wrapper.state('subSessions')).toEqual([]);
    expect(minProps.triggerResize).toBeCalled();
    expect(wrapper).toMatchSnapshot();
  });

  test('Rule Description overlay state is set to true and render correctly when setOverlay is invoked', () => {
    wrapper
      .find('Description')
      .at(0)
      .prop('setOverlay')();

    expect(wrapper.state('overlay')).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  test('Rule Description overlay state is set to false and render correctly when removeOverlay is invoked', () => {
    wrapper
      .find('Description')
      .at(0)
      .prop('removeOverlay')();

    expect(wrapper.state('overlay')).toBe(false);
    expect(minProps.clearErrorNotification).toBeCalled();
    expect(wrapper).toMatchSnapshot();
  });

  test('handleParamSet prop is called with correct paramaters when handleParamSet is invoked', () => {
    wrapper
      .find('Description')
      .at(0)
      .prop('handleParamSet')(
      ruleDescriptionData[0].userDescription.references[0].value,
      ruleDescriptionData[0].userDescription,
      rulesData[0].code
    );

    expect(minProps.handleParamSet).toBeCalledWith(
      ruleDescriptionData[0].userDescription.references[0].value,
      ruleDescriptionData[0].userDescription,
      rulesData[0].code
    );
  });

  test('handleParamReset prop is called with correct paramaters when handleParamReset is invoked', () => {
    wrapper
      .find('Description')
      .at(0)
      .prop('handleParamReset')(ruleDescriptionData[0].userDescription);

    expect(minProps.handleParamReset).toBeCalledWith(
      ruleDescriptionData[0].userDescription
    );
  });

  test('Check whether reportError prop is called', () => {
    wrapper
      .find('Description')
      .at(0)
      .prop('reportError')({ error: 'some error' });

    expect(minProps.reportError).toBeCalledWith({
      error: { error: 'some error' },
    });
  });
});
