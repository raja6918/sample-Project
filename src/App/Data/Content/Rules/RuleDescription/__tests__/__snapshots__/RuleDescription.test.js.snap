// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Rule Description Component Rule Description Component renders correctly 1`] = `
<div
  className="rule-description"
>
  <RuleDescription__PrimarySessionDiv>
    <Description
      code="CiesApplyAllowDeadheadDuty"
      data={
        Object {
          "data": "Allow deadhead-only duties in a pairing at the beginning, middle or end of a pairing, according to the following selections.
[SierraCiesApplyAllowDeadheadDuty0]() Allow deadhead-only duties at the beginning of a pairing.
[SierraCiesApplyAllowDeadheadDuty1]() Allow deadhead-only duties anywhere in the middle of a pairing.
[SierraCiesApplyAllowDeadheadDuty2]() Allow deadhead-only duties at the end of a pairing.
Note: If you want to forbid a deadhead-only duty, remove that duty type from the [hyperlink0]().",
          "object": "userDescriptionComponent",
          "references": Array [
            Object {
              "definedInCurrent": false,
              "description": "",
              "name": "SierraCiesApplyAllowDeadheadDuty0",
              "object": "userDescriptionParamItem",
              "paramType": Object {
                "object": "booleanParamType",
              },
              "referenceRulesetName": "root",
              "switch": true,
              "type": "booleanParamType",
              "value": "true",
            },
            Object {
              "definedInCurrent": false,
              "description": "",
              "name": "SierraCiesApplyAllowDeadheadDuty1",
              "object": "userDescriptionParamItem",
              "paramType": Object {
                "object": "booleanParamType",
              },
              "referenceRulesetName": "root",
              "switch": true,
              "type": "booleanParamType",
              "value": "true",
            },
            Object {
              "definedInCurrent": false,
              "description": "",
              "name": "SierraCiesApplyAllowDeadheadDuty2",
              "object": "userDescriptionParamItem",
              "paramType": Object {
                "object": "booleanParamType",
              },
              "referenceRulesetName": "root",
              "switch": true,
              "type": "booleanParamType",
              "value": "true",
            },
            Object {
              "code": "CiesApplyDutyComposition",
              "name": "hyperlink0",
              "object": "userDescriptionLinkItem",
              "type": "hyperlink",
              "value": "Duty Composition rule",
            },
          ],
        }
      }
      handleLink={[Function]}
      handleParamReset={[MockFunction]}
      handleParamSet={[MockFunction]}
      overlay={false}
      readOnly={false}
      refreshRules={[MockFunction]}
      removeOverlay={[Function]}
      reportError={[Function]}
      ruleSet={2}
      scenarioId={1}
      setOverlay={[Function]}
      t={[MockFunction]}
      toggleLoader={[MockFunction]}
      userId={1}
    />
  </RuleDescription__PrimarySessionDiv>
</div>
`;

exports[`Test Rule Description Component Rule Description overlay state is set to false and render correctly when removeOverlay is invoked 1`] = `
<div
  className="rule-description"
>
  <RuleDescription__PrimarySessionDiv>
    <Description
      code="CiesApplyAllowDeadheadDuty"
      data={
        Object {
          "data": "Allow deadhead-only duties in a pairing at the beginning, middle or end of a pairing, according to the following selections.
[SierraCiesApplyAllowDeadheadDuty0]() Allow deadhead-only duties at the beginning of a pairing.
[SierraCiesApplyAllowDeadheadDuty1]() Allow deadhead-only duties anywhere in the middle of a pairing.
[SierraCiesApplyAllowDeadheadDuty2]() Allow deadhead-only duties at the end of a pairing.
Note: If you want to forbid a deadhead-only duty, remove that duty type from the [hyperlink0]().",
          "object": "userDescriptionComponent",
          "references": Array [
            Object {
              "definedInCurrent": false,
              "description": "",
              "name": "SierraCiesApplyAllowDeadheadDuty0",
              "object": "userDescriptionParamItem",
              "paramType": Object {
                "object": "booleanParamType",
              },
              "referenceRulesetName": "root",
              "switch": true,
              "type": "booleanParamType",
              "value": "true",
            },
            Object {
              "definedInCurrent": false,
              "description": "",
              "name": "SierraCiesApplyAllowDeadheadDuty1",
              "object": "userDescriptionParamItem",
              "paramType": Object {
                "object": "booleanParamType",
              },
              "referenceRulesetName": "root",
              "switch": true,
              "type": "booleanParamType",
              "value": "true",
            },
            Object {
              "definedInCurrent": false,
              "description": "",
              "name": "SierraCiesApplyAllowDeadheadDuty2",
              "object": "userDescriptionParamItem",
              "paramType": Object {
                "object": "booleanParamType",
              },
              "referenceRulesetName": "root",
              "switch": true,
              "type": "booleanParamType",
              "value": "true",
            },
            Object {
              "code": "CiesApplyDutyComposition",
              "name": "hyperlink0",
              "object": "userDescriptionLinkItem",
              "type": "hyperlink",
              "value": "Duty Composition rule",
            },
          ],
        }
      }
      handleLink={[Function]}
      handleParamReset={[MockFunction]}
      handleParamSet={[MockFunction]}
      overlay={false}
      readOnly={false}
      refreshRules={[MockFunction]}
      removeOverlay={[Function]}
      reportError={[Function]}
      ruleSet={2}
      scenarioId={1}
      setOverlay={[Function]}
      t={[MockFunction]}
      toggleLoader={[MockFunction]}
      userId={1}
    />
  </RuleDescription__PrimarySessionDiv>
</div>
`;

exports[`Test Rule Description Component Rule Description overlay state is set to true and render correctly when setOverlay is invoked 1`] = `
<div
  className="rule-description"
>
  <div
    className="overlay"
    onClick={[Function]}
  />
  <RuleDescription__PrimarySessionDiv>
    <Description
      code="CiesApplyAllowDeadheadDuty"
      data={
        Object {
          "data": "Allow deadhead-only duties in a pairing at the beginning, middle or end of a pairing, according to the following selections.
[SierraCiesApplyAllowDeadheadDuty0]() Allow deadhead-only duties at the beginning of a pairing.
[SierraCiesApplyAllowDeadheadDuty1]() Allow deadhead-only duties anywhere in the middle of a pairing.
[SierraCiesApplyAllowDeadheadDuty2]() Allow deadhead-only duties at the end of a pairing.
Note: If you want to forbid a deadhead-only duty, remove that duty type from the [hyperlink0]().",
          "object": "userDescriptionComponent",
          "references": Array [
            Object {
              "definedInCurrent": false,
              "description": "",
              "name": "SierraCiesApplyAllowDeadheadDuty0",
              "object": "userDescriptionParamItem",
              "paramType": Object {
                "object": "booleanParamType",
              },
              "referenceRulesetName": "root",
              "switch": true,
              "type": "booleanParamType",
              "value": "true",
            },
            Object {
              "definedInCurrent": false,
              "description": "",
              "name": "SierraCiesApplyAllowDeadheadDuty1",
              "object": "userDescriptionParamItem",
              "paramType": Object {
                "object": "booleanParamType",
              },
              "referenceRulesetName": "root",
              "switch": true,
              "type": "booleanParamType",
              "value": "true",
            },
            Object {
              "definedInCurrent": false,
              "description": "",
              "name": "SierraCiesApplyAllowDeadheadDuty2",
              "object": "userDescriptionParamItem",
              "paramType": Object {
                "object": "booleanParamType",
              },
              "referenceRulesetName": "root",
              "switch": true,
              "type": "booleanParamType",
              "value": "true",
            },
            Object {
              "code": "CiesApplyDutyComposition",
              "name": "hyperlink0",
              "object": "userDescriptionLinkItem",
              "type": "hyperlink",
              "value": "Duty Composition rule",
            },
          ],
        }
      }
      handleLink={[Function]}
      handleParamReset={[MockFunction]}
      handleParamSet={[MockFunction]}
      overlay={true}
      readOnly={false}
      refreshRules={[MockFunction]}
      removeOverlay={[Function]}
      reportError={[Function]}
      ruleSet={2}
      scenarioId={1}
      setOverlay={[Function]}
      t={[MockFunction]}
      toggleLoader={[MockFunction]}
      userId={1}
    />
  </RuleDescription__PrimarySessionDiv>
</div>
`;

exports[`Test Rule Description Component Rule Description subSessions state and subsession renders correctly when handleLink is invoked 1`] = `
<div
  className="rule-description"
>
  <RuleDescription__PrimarySessionDiv>
    <Description
      code="CiesApplyAllowDeadheadDuty"
      data={
        Object {
          "data": "Allow deadhead-only duties in a pairing at the beginning, middle or end of a pairing, according to the following selections.
[SierraCiesApplyAllowDeadheadDuty0]() Allow deadhead-only duties at the beginning of a pairing.
[SierraCiesApplyAllowDeadheadDuty1]() Allow deadhead-only duties anywhere in the middle of a pairing.
[SierraCiesApplyAllowDeadheadDuty2]() Allow deadhead-only duties at the end of a pairing.
Note: If you want to forbid a deadhead-only duty, remove that duty type from the [hyperlink0]().",
          "object": "userDescriptionComponent",
          "references": Array [
            Object {
              "definedInCurrent": false,
              "description": "",
              "name": "SierraCiesApplyAllowDeadheadDuty0",
              "object": "userDescriptionParamItem",
              "paramType": Object {
                "object": "booleanParamType",
              },
              "referenceRulesetName": "root",
              "switch": true,
              "type": "booleanParamType",
              "value": "true",
            },
            Object {
              "definedInCurrent": false,
              "description": "",
              "name": "SierraCiesApplyAllowDeadheadDuty1",
              "object": "userDescriptionParamItem",
              "paramType": Object {
                "object": "booleanParamType",
              },
              "referenceRulesetName": "root",
              "switch": true,
              "type": "booleanParamType",
              "value": "true",
            },
            Object {
              "definedInCurrent": false,
              "description": "",
              "name": "SierraCiesApplyAllowDeadheadDuty2",
              "object": "userDescriptionParamItem",
              "paramType": Object {
                "object": "booleanParamType",
              },
              "referenceRulesetName": "root",
              "switch": true,
              "type": "booleanParamType",
              "value": "true",
            },
            Object {
              "code": "CiesApplyDutyComposition",
              "name": "hyperlink0",
              "object": "userDescriptionLinkItem",
              "type": "hyperlink",
              "value": "Duty Composition rule",
            },
          ],
        }
      }
      handleLink={[Function]}
      handleParamReset={[MockFunction]}
      handleParamSet={[MockFunction]}
      overlay={false}
      readOnly={false}
      refreshRules={[MockFunction]}
      removeOverlay={[Function]}
      reportError={[Function]}
      ruleSet={2}
      scenarioId={1}
      setOverlay={[Function]}
      t={[MockFunction]}
      toggleLoader={[MockFunction]}
      userId={1}
    />
  </RuleDescription__PrimarySessionDiv>
  <RuleDescription__SecondarySessionDiv>
    <RuleDescription__BorderDiv />
    <RuleDescription__SubSessionDiv
      key="CiesApplyAllowDeadheadDuty"
    >
      <span>
        <RuleDescription__SubSessionsTitle
          style={
            Object {
              "paddingBottom": "",
            }
          }
        >
          Allow deadhead duties at the start, middle, and/or end of a pairing
        </RuleDescription__SubSessionsTitle>
        <Description
          code="CiesApplyAllowDeadheadDuty"
          data={
            Object {
              "data": "Allow deadhead-only duties in a pairing at the beginning, middle or end of a pairing, according to the following selections.
[SierraCiesApplyAllowDeadheadDuty0]() Allow deadhead-only duties at the beginning of a pairing.
[SierraCiesApplyAllowDeadheadDuty1]() Allow deadhead-only duties anywhere in the middle of a pairing.
[SierraCiesApplyAllowDeadheadDuty2]() Allow deadhead-only duties at the end of a pairing.
Note: If you want to forbid a deadhead-only duty, remove that duty type from the [hyperlink0]().",
              "object": "userDescriptionComponent",
              "references": Array [
                Object {
                  "definedInCurrent": false,
                  "description": "",
                  "name": "SierraCiesApplyAllowDeadheadDuty0",
                  "object": "userDescriptionParamItem",
                  "paramType": Object {
                    "object": "booleanParamType",
                  },
                  "referenceRulesetName": "root",
                  "switch": true,
                  "type": "booleanParamType",
                  "value": "true",
                },
                Object {
                  "definedInCurrent": false,
                  "description": "",
                  "name": "SierraCiesApplyAllowDeadheadDuty1",
                  "object": "userDescriptionParamItem",
                  "paramType": Object {
                    "object": "booleanParamType",
                  },
                  "referenceRulesetName": "root",
                  "switch": true,
                  "type": "booleanParamType",
                  "value": "true",
                },
                Object {
                  "definedInCurrent": false,
                  "description": "",
                  "name": "SierraCiesApplyAllowDeadheadDuty2",
                  "object": "userDescriptionParamItem",
                  "paramType": Object {
                    "object": "booleanParamType",
                  },
                  "referenceRulesetName": "root",
                  "switch": true,
                  "type": "booleanParamType",
                  "value": "true",
                },
                Object {
                  "code": "CiesApplyDutyComposition",
                  "name": "hyperlink0",
                  "object": "userDescriptionLinkItem",
                  "type": "hyperlink",
                  "value": "Duty Composition rule",
                },
              ],
            }
          }
          handleLink={[Function]}
          handleParamReset={[MockFunction]}
          handleParamSet={[MockFunction]}
          overlay={false}
          readOnly={false}
          refreshRules={[MockFunction]}
          removeOverlay={[Function]}
          reportError={[Function]}
          ruleSet={2}
          scenarioId={1}
          setOverlay={[Function]}
          t={[MockFunction]}
          toggleLoader={[MockFunction]}
          userId={1}
        />
      </span>
      <span
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <WithStyles(ForwardRef(IconButton))
          aria-label="close row"
          onClick={[Function]}
          size="small"
          style={
            Object {
              "float": "right",
            }
          }
        >
          <Memo([object Object])
            color="disabled"
            fontSize="small"
          />
        </WithStyles(ForwardRef(IconButton))>
      </span>
    </RuleDescription__SubSessionDiv>
  </RuleDescription__SecondarySessionDiv>
</div>
`;

exports[`Test Rule Description Component Rule Description subSessions state reset and subsession renders correctly when handleClose is invoked 1`] = `
<div
  className="rule-description"
>
  <RuleDescription__PrimarySessionDiv>
    <Description
      code="CiesApplyAllowDeadheadDuty"
      data={
        Object {
          "data": "Allow deadhead-only duties in a pairing at the beginning, middle or end of a pairing, according to the following selections.
[SierraCiesApplyAllowDeadheadDuty0]() Allow deadhead-only duties at the beginning of a pairing.
[SierraCiesApplyAllowDeadheadDuty1]() Allow deadhead-only duties anywhere in the middle of a pairing.
[SierraCiesApplyAllowDeadheadDuty2]() Allow deadhead-only duties at the end of a pairing.
Note: If you want to forbid a deadhead-only duty, remove that duty type from the [hyperlink0]().",
          "object": "userDescriptionComponent",
          "references": Array [
            Object {
              "definedInCurrent": false,
              "description": "",
              "name": "SierraCiesApplyAllowDeadheadDuty0",
              "object": "userDescriptionParamItem",
              "paramType": Object {
                "object": "booleanParamType",
              },
              "referenceRulesetName": "root",
              "switch": true,
              "type": "booleanParamType",
              "value": "true",
            },
            Object {
              "definedInCurrent": false,
              "description": "",
              "name": "SierraCiesApplyAllowDeadheadDuty1",
              "object": "userDescriptionParamItem",
              "paramType": Object {
                "object": "booleanParamType",
              },
              "referenceRulesetName": "root",
              "switch": true,
              "type": "booleanParamType",
              "value": "true",
            },
            Object {
              "definedInCurrent": false,
              "description": "",
              "name": "SierraCiesApplyAllowDeadheadDuty2",
              "object": "userDescriptionParamItem",
              "paramType": Object {
                "object": "booleanParamType",
              },
              "referenceRulesetName": "root",
              "switch": true,
              "type": "booleanParamType",
              "value": "true",
            },
            Object {
              "code": "CiesApplyDutyComposition",
              "name": "hyperlink0",
              "object": "userDescriptionLinkItem",
              "type": "hyperlink",
              "value": "Duty Composition rule",
            },
          ],
        }
      }
      handleLink={[Function]}
      handleParamReset={[MockFunction]}
      handleParamSet={[MockFunction]}
      overlay={false}
      readOnly={false}
      refreshRules={[MockFunction]}
      removeOverlay={[Function]}
      reportError={[Function]}
      ruleSet={2}
      scenarioId={1}
      setOverlay={[Function]}
      t={[MockFunction]}
      toggleLoader={[MockFunction]}
      userId={1}
    />
  </RuleDescription__PrimarySessionDiv>
</div>
`;
